### PR TITLE
Add streaming ASR with Emformer RNN-T

### PR DESCRIPTION
**Pull request by csukuangfj**
I have tested it and it works. Will upload a pretrained Emformer model later.

Note that the framework is quite general and it is easy to adapt to other kinds of stateless RNN-T models, not limited to Emformer RNN-T models.

### CODE REVIEW

**Comment by pkufool** [comment]
Wow, Cool! Once you merge this PR, I will try to add streaming conformer model.

File: sherpa/bin/decode.py [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+
+import torch
+from kaldifeat import FbankOptions, OnlineFbank, OnlineFeature
+
+
+def unstack_states(
+    states: List[List[torch.Tensor]],
+) -> List[List[List[torch.Tensor]]]:
+    """Unstack the Emformer state corresponding to a batch of utterances
+    into a list of states, wwere the i-th entry is the state from the i-th
```
Comment (pkufool) [inline sherpa/bin/decode.py:28]: wwere -> where

File: sherpa/bin/streaming_server.py [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+        self.initial_decoder_out = self.model.decoder_forward(decoder_input).squeeze(1)
+
+        self.nn_pool = ThreadPoolExecutor(
+            max_workers=2,
+            thread_name_prefix="nn",
+        )
+
+        self.stream_queue = asyncio.Queue()
+        self.max_wait_ms = 10
+        self.max_batch_size = 20
```
Comment (pkufool) [inline sherpa/bin/streaming_server.py:198]: I think it is better to make `max_batch_size` configurable.

File: sherpa/csrc/rnnt_emformer_model.h [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "torch/script.h"
+
+namespace sherpa {
+
+/** It wraps a torch script model, which is from
+ * pruned_stateless_emformer_rnnt/mode.py within icefall.
```
Comment (pkufool) [inline sherpa/csrc/rnnt_emformer_model.h:30]: mode -> model

File: sherpa/csrc/rnnt_model.h [hunk 4]
```diff
@@ -53,7 +53,7 @@ class RnntModel {
    * @param features  A 3-D tensor of shape (N, T, C).
    * @param features_length A 1-D tensor of shape (N,) containing the number of
    *                       valid frames in `features`.
-   * @return Return a tuple containing two tensors:
+   * @return Return a pair containing two tensors:
```
Comment (pkufool) [inline sherpa/csrc/rnnt_model.h:56]: I think we will need `feature_lens` in `fast_beam_search`. But surely we might add it when needed.

**Review by pkufool — APPROVED** [review]
LGTM

File: sherpa/csrc/rnnt_beam_search.cc [hunk 5]
```diff
@@ -141,4 +146,50 @@ std::vector<std::vector<int32_t>> GreedySearch(
...
+        (*hyps)[n].push_back(index);
+      }
+    }
+
+    if (emitted) {
+      BuildDecoderInput(*hyps, &decoder_input);
+      decoder_out = model.ForwardDecoder(decoder_input.to(device)).squeeze(1);
+    }
+  }
+  return decoder_out;
```
[the remark this answers is not in the document]
Comment (csukuangfj) [reply sherpa/csrc/rnnt_beam_search.cc:192]: `decoder_out` is to save an extra op `decoder.forward()`.

For streaming decoding, the input chunk size is fixed and there are no paddings. We can figure out the encoder_out_len from encoder_out.

We can add it for fast_beam_search later if it turns out it is necesseary.

**Comment by csukuangfj** [comment]
Here is a demo for this PR: https://www.youtube.com/watch?v=z7HgaZv5W0U

<a href="http://www.youtube.com/watch?feature=player_embedded&v=z7HgaZv5W0U" target="_blank">
 Watch the video
</a>

**Pull request merged**